yarn/history/src/main/scala/org/apache/spark/deploy/history/yarn/YarnHistoryProvider.scala

```diff
@@ -0,0 +1,1015 @@
+        setLastFailure(results.failureCause.get, results.timestamp)
+        // and the failure counter
+        refreshFailedCount.incrementAndGet()
+      }
+    }
+    results
+  }
+
+  /**
+   * List applications. This currently finds completed applications only.
+   * 
+   * If the timeline is  not enabled, returns an empty list
+   * @return List of all known applications.
+   */
+  override def getListing(): Seq[ApplicationHistoryInfo] = {
+    // get the current list
+    val listing = getApplications().applications
+    // and queue another refresh
+    triggerRefresh()
+    listing
+  }
+
+  /**
+   * Trigger a refresh
+   */
+  def triggerRefresh(): Unit = {
+    refresher.refresh(now())
+  }
+
+  /**
+   * Return the current time
+   * @return
+   */
+  def now(): Long = {
+    System.currentTimeMillis()
+  }
+
+  /**
+   * Get the last refresh attempt (Which may or may not be successful)
+   * @return the last refresh time
+   */
+  def getLastRefreshAttemptTime(): Long = {
+    refresher.lastRefreshAttemptTime
+  }
+  
+  /**
+   * Look up the timeline entity
+   * @param appId application ID
+   * @return the entity associated with the given application
+   * @throws FileNotFoundException if no entry was found
+   */
+  def getTimelineEntity(appId: String): TimelineEntity = {
+    logDebug(s"GetTimelineEntity $appId")
+    maybeCheckHealth()
+    getTimelineQueryClient().getEntity(YarnHistoryService.SPARK_EVENT_ENTITY_TYPE, appId)
+  }
+
+
+  /**
+   * Returns the Spark UI for a specific application.
+   * <p>
+   * If the timeline is  not enabled, returns `None`
+   * @param appId The application ID.
+   * @param attemptId The application attempt ID (or `None` if there is no attempt ID).
+   * @return The application's UI, or `None` if application is not found.
+   */
+  override def getAppUI(appId: String, attemptId: Option[String]): Option[SparkUI] = {
+    getAppUI(appId)
+  }
+
+  /**
+   * Build the application UI for an application
+   * <p>
+   * If the timeline is  not enabled, returns `None`
+   * @param appId The application ID.
+   * @return The application's UI, or `None` if application is not found.
+   */
+  def getAppUI(appId: String): Option[SparkUI] = {
+    logDebug(s"Request UI with appId $appId")
+    if (!enabled) {
+      // Timeline is disabled: return nothing
+      return None
+    }
+    maybeCheckHealth()
+    try {
+      val entity = getTimelineEntity(appId)
+
+      if (log.isDebugEnabled) {
+        logDebug(describeEntity(entity))
+      }
+      val bus = new SparkListenerBus() {}
+      val appListener = new ApplicationEventListener()
+      bus.addListener(appListener)
+
+      val ui = {
+        val conf = this.sparkConf.clone()
+        val appSecManager = new SecurityManager(conf)
+        SparkUI.createHistoryUI(conf, bus, appSecManager, appId,
+                                 HistoryServer.UI_PATH_PREFIX + s"/${appId }", entity.getStartTime)
+      }
+      val events = entity.getEvents
```

Comment:
> If you ever look at where CPU load

I'm not worried about CPU load. I'm worried about a large number of events using a lot of memory both in the ATS and in the history server. On the ATS side, you have to materialize all the events into a list in memory and then serialize them to JSON in memory before writing to the socket. Similarly, on the HS you need to read all that serialized JSON into memory and then materialize it into deserialized objects. 

For an application with a really large number of events, that's a very non-trivial amount of memory. This is irrelevant to whether the app is complete or not.